tests/test_contracts.py

```diff
@@ -202,3 +203,36 @@ def test_errback(self):
 
         self.assertFalse(self.results.failures)
         self.assertTrue(self.results.errors)
+
+    @defer.inlineCallbacks
+    def test_same_url(self):
+
+        class TestSameUrlSpider(Spider):
+            name = 'test_same_url'
+
+            def __init__(self, *args, **kwargs):
+                super(TestSameUrlSpider, self).__init__(*args, **kwargs)
+                self.visited = 0
+
+            def start_requests(s):
+                return self.conman.from_spider(s, self.results)
+
+            def parse_first(self, response):
+                """first callback
+                @url http://scrapy.org
+                """
+                self.visited += 1
+                return TestItem()
+
+            def parse_second(self, response):
+                """second callback
+                @url http://scrapy.org
+                """
+                self.visited += 1
+                return TestItem()
+
+        crawler = CrawlerRunner().create_crawler(TestSameUrlSpider)
+        with MockServer() as mockserver:
+            yield crawler.crawl(mockserver=mockserver)
```

Comment:
Unfortunately mockserver doesn't make Scrapy magically make local requests :) MockServer just starts a process which listens to HTTP, and has a few endpoints you can use for testing. To actually use it you need to instruct Scrapy to make requests to it. 

In Scrapy Contracts URL to fetch is defined in a docstring (`@url ...`), so if you define `http://scrapy.org` there, this test will fetch scrapy.org website. A sanity check is to try running this test with Internet disabled (e.g. WiFi turned off) - Scrapy testing suite shouldn't need external resources to run. To make it work you need to use an URL of some mockserver endpoint instead of scrapy.org.

Reply:
crawler.crawl arguments are passed to `Spider.__init__`, which eventually sets them as attributes. So this line does the following: starts crawling using TestSameUrlSpider, and this spider has `mockserver` attribute set to `mockserver` instance - which is not useful, as you're not using it in a spider.